spec/haproxy/templates/haproxy_config/frontend_https_spec.rb

```diff
@@ -178,68 +161,184 @@
+        expect(frontend_https).to include('http-request del-header X-SSL-Client-NotBefore  if !route_service_request')
+        expect(frontend_https).to include('http-request del-header X-SSL-Client-NotAfter   if !route_service_request')
       end
-    end
-  end
 
-  context 'when ha_proxy.forwarded_client_cert is forward_only_if_route_service' do
-    let(:properties) do
-      default_properties.merge({ 'forwarded_client_cert' => 'forward_only_if_route_service' })
-    end
+      it 'does not add mTLS headers' do
+        expect(frontend_https).not_to include(/http-request set-header X-Fowarded-Client-Cert/)
+        expect(frontend_https).not_to include(/http-request set-header X-SSL-Client/)
+      end
+
+      context 'when mutual TLS is enabled' do
+        let(:properties) do
+          default_properties.merge({
+            'client_cert' => true,
+            'forwarded_client_cert' => 'forward_only_if_route_service'
+          })
+        end
+
+        it 'deletes mTLS headers for non-route service requests (for mTLS and non-mTLS)' do
+          expect(frontend_https).to include('acl route_service_request hdr(X-Cf-Proxy-Signature) -m found')
+          expect(frontend_https).to include('http-request del-header X-Forwarded-Client-Cert if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client            if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-Session-ID if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-Verify     if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-Subject-DN if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-Subject-CN if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-Issuer-DN  if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-NotBefore  if !route_service_request')
+          expect(frontend_https).to include('http-request del-header X-SSL-Client-NotAfter   if !route_service_request')
+        end
 
-    it 'deletes the X-Forwarded-Client-Cert header for non-route service requests' do
-      expect(frontend_https).to include('acl route_service_request hdr(X-Cf-Proxy-Signature) -m found')
-      expect(frontend_https).to include('http-request del-header X-Forwarded-Client-Cert if !route_service_request')
-      expect(frontend_https).to include('http-request set-header X-Forwarded-Client-Cert %[ssl_c_der,base64] if { ssl_c_used }')
+        it 'overwrites mTLS headers when mTLS is used' do
+          expect(frontend_https).to include('http-request set-header X-Forwarded-Client-Cert %[ssl_c_der,base64]      if { ssl_c_used }')
```

Comment:
The new solution is a lot easier to understand, although there is one downside to it: Previous solutions only used `set-header` which implies overwriting previous instances so a `del-header` was not necessary. I still like the new way better, even if it is more verbose.

Reply:
Are you sure? There are del-headers here that happen regardless 

https://github.com/cloudfoundry-incubator/haproxy-boshrelease/blob/master/jobs/haproxy/templates/haproxy.config.erb#L361

https://github.com/cloudfoundry-incubator/haproxy-boshrelease/blob/master/jobs/haproxy/templates/haproxy.config.erb#L363

https://github.com/cloudfoundry-incubator/haproxy-boshrelease/blob/master/jobs/haproxy/templates/haproxy.config.erb#L369

Reply:
I'm not saying the old logic was great :) Only that "set-header" implies "del-header" + "add-header" but it really is easier to read now. 

Docs: http://cbonte.github.io/haproxy-dconv/2.2/configuration.html#4.2-http-request%20set-header

Reply:
Hi @domdom82 I've tried not to change any logic for `X-Forwarded-Client-Cert` -- if I have changed it then that is a mistake :) . There were already `del-header`s in the original implementation which is why I've kept them. You're right that set-header overrides del-header, but the set-headers are conditional (eg `if { ssl_c_used }`) which is why I assume the del-headers were added originally.